### PR TITLE
feat(partition): stricter types, add test

### PR DESCRIPTION
**Pull request by GerkinDev**

### CODE REVIEW

File: types/partition.d.ts [hunk 1]
```diff
@@ -1,4 +1,8 @@
...
-export function partition<T>(fn: (a: T) => boolean): (list: readonly T[]) => [T[], T[]];
-export function partition(fn: (a: string) => boolean, list: readonly string[]): [string[], string[]];
+export function partition<T, U extends T>(fn: (a: T) => a is U, list: readonly T[]): [U[], Exclude<T, U>[]];
 export function partition<T>(fn: (a: T) => boolean, list: readonly T[]): [T[], T[]];
+
+export function partition<T, U extends T>(fn: (a: T) => a is U): {
+  <L extends T = T>(list: readonly L[]): [(L & U)[], Exclude<L, U>[]];
+  (list: readonly T[]): [(T & U)[], Exclude<T, U>[]];
+};
+export function partition<T>(fn: (a: T) => boolean): <L extends T = T>(list: readonly L[]) => [L[], L[]];
```
Comment (Harris-Miller) [inline types/partition.d.ts:8]: The `fn: (a: T) => a is U` support is a great addition.

Thank you for removing the `string` specific overload as well. That was copy/pasted from the original definition over in `@types/ramda`. I'm not sure why it was needed as the generic overload covers that case as well

Couple things:

The order of the overloads is important, you need to have `partition(fn)(list)` defined before `partition(fn, list)`. When passing `partition` to another function, eg `flip(partition)`, typescript asumes the _last_ overload as the most general and uses that


```typescript
export function partition<T, U extends T>(fn: (a: T) => a is U): {
  <L extends T = T>(list: readonly L[]): [(L & U)[], Exclude<L, U>[]];
  (list: readonly T[]): [(T & U)[], Exclude<T, U>[]];
};
```
I don't understand the 2 overloads in the middle

* `<L extends T = T>(list: readonly L[]): [(L & U)[], Exclude<L, U>[]];`
  * What is the generic `L` accomplishing here?
* `(list: readonly T[]): [(T & U)[], Exclude<T, U>[]];`
  * The intersection `T & U` seems extranious. `U` is already defied to extend `T`, and therefore is a subset of `T`. So by definition `T & U === T`.

File: test/partition.test.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+// fn: (v) => v is T
+expectType<[true[], false[]]>(partition((v): v is true => v, [true, false]));
+// fn: (v) => bool
+expectType<[boolean[], boolean[]]>(partition((v) => v, [true, false]));
+
+// partition(fn)(list)
+// fn: (v) => v is T
+expectType<[true[], false[]]>(partition((v): v is true => !!v)([true, false]));
+// fn: (v) => bool
+expectType<[boolean[], boolean[]]>(partition((v) => !!v)([true, false]));
```
Comment (Harris-Miller) [inline test/partition.test.ts:14]: In my other comment, the bits about `L extends T = T` and `T & U`. If those are necessary for reasons I'm just not aware of, the best way to show that would be through tests. If you can create tests for the specific situation that would require those and would fail otherwise, that would help greatly

**Comment by Harris-Miller** [comment]
While testing this, I discovered a very interesting edge case for `isNil`: https://tsplay.dev/m3Kpjm

For `partition` to work correctly with `isNil`, that will need to be fixed first: https://github.com/ramda/types/pull/90

File: types/partition.d.ts (hunk 1 above)
Comment (GerkinDev) [reply types/partition.d.ts:8]: - the first overload is required for this test: https://github.com/ramda/types/blob/8e6cae9adbe097d854927b44dbaaf0cc4dde9dc2/test/partition.test.ts#L12 I also added another test with list type resolved at call-time here: https://github.com/ramda/types/blob/8e6cae9adbe097d854927b44dbaaf0cc4dde9dc2/test/partition.test.ts#L15-L17
- I removed the useless intersection

Comment (Harris-Miller) [reply types/partition.d.ts:8]: Ah ok, the `<L extends T= T>` generic makes sense to me now, here is an example that uses an `isNumber` function: https://tsplay.dev/mpz56w

We actually don't need both of these then:
```typescript
  <L extends T = T>(list: readonly L[]): [(L & U)[], Exclude<L, U>[]];
  (list: readonly T[]): [(T & U)[], Exclude<T, U>[]];
```
Just have it be
```typescript
  <L extends T = T>(list: readonly L[]): [(L & U)[], Exclude<L, U>[]];
```

I'm glad to know about this trick, there should be many other places where we can apply it

Comment (Harris-Miller) [reply types/partition.d.ts:8]: Turns out that's how `filter` works, so at least we're consistent now!

Comment (GerkinDev) [reply types/partition.d.ts:8]: FYI, types of rxjs, which have a great implementation of a `pipe` pattern, are almost all functions retuning generics just like this

Comment (GerkinDev) [reply types/partition.d.ts:5]: I've removed the useless overload

**Comment by GerkinDev** [comment]
What do you need from me to consider this PR as mergeable ?

**Comment by Harris-Miller** [comment]
> What do you need from me to consider this PR as mergeable?

@GerkinDev I merged https://github.com/ramda/types/pull/90, so

* Rebase on `develop`
* Re-order the overloads
  * The full signature always needs to come last

File: types/partition.d.ts (hunk 1 above)
Comment (Harris-Miller) [reply types/partition.d.ts:5]: re: `pipe`
rxjs: https://github.com/ReactiveX/rxjs/blob/c66657469dc17fbed992e273d7d121e51c00d671/packages/rxjs/src/internal/util/pipe.ts
ramda: https://github.com/ramda/types/blob/develop/types/pipe.d.ts

They're very close

Comment (GerkinDev) [reply types/partition.d.ts:5]: For pipe, yes, but I was thinking about operators

Comment (Harris-Miller) [reply types/partition.d.ts:4]: Which operators?

Comment (Harris-Miller) [reply types/partition.d.ts:4]: Sorry one last thing, the overloads, both `(fn)(arr)`s need to be before `(fn, arr)`
so this:
```typescript
partition(fn: val is T)(arr)
partition(fn: boolean)(arr)
partition(fn: val is T, arr)
partition(fn: boolean, arr)
```
The rule is "specific first, general last", and while it doesn't exactly explain it here: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#function-overloads, that does include for currying

**Review by Harris-Miller — DISMISSED** [review]

File: types/partition.d.ts (hunk 1 above)
Comment (GerkinDev) [reply types/partition.d.ts:4]: [partition](https://github.com/ReactiveX/rxjs/blob/c66657469dc17fbed992e273d7d121e51c00d671/packages/rxjs/src/internal/operators/partition.ts), [map](https://github.com/ReactiveX/rxjs/blob/c66657469dc17fbed992e273d7d121e51c00d671/packages/rxjs/src/internal/operators/map.ts), [filter](https://github.com/ReactiveX/rxjs/blob/c66657469dc17fbed992e273d7d121e51c00d671/packages/rxjs/src/internal/operators/filter.ts), etc. Almost all of them returns generic unbound unary functions

**Review by Harris-Miller — APPROVED** [review]
Looks good
